### PR TITLE
[Draft][Parser] Enable the new text parser by default

### PR DESCRIPTION
**Pull request by tlively**

### CODE REVIEW

**Comment by tlively** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6371?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* **#6371** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6371?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#6370** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6370?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/6371?utm_source=stack-comment).

File: test/lit/debug/full.wat [hunk 1]
```diff
@@ -20,33 +20,26 @@
   ;; NRML-NEXT:   (drop
   ;; NRML-NEXT:    (i32.const 1)
   ;; NRML-NEXT:   )
-  ;; NRML-NEXT:   ;;@ src.cpp:3:4
```
Comment (tlively) [inline test/lit/debug/full.wat:23]: Looks like I misunderstood some of the source location stuff :)

Is it the case that a source location annotation applies to every following instruction in stack machine order until there's a new source location annotation? Do they continue applying beyond the beginnings and ends of control flow scopes?

File: test/lit/multivalue.wast [hunk 2]
```diff
@@ -356,10 +356,24 @@
 
  ;; Test multivalue control structures
  ;; CHECK:      (func $mv-return (type $0) (result i32 i64)
+ ;; CHECK-NEXT:  (local $scratch i32)
```
Comment (tlively) [inline test/lit/multivalue.wast:359]: This return lowering can be fixed in IRBuilder.

File: test/lit/debug/full.wat (hunk 1 above)
Comment (kripken) [reply test/lit/debug/full.wat:23]: I think you may be referring to the issue fixed in #5906, see details there.

Comment (tlively) [reply test/lit/debug/full.wat:23]: Hmm, no, that PR seems to only affect the encoding of source map info, but I'm confused about how to properly parser source map info.

Comment (kripken) [reply test/lit/debug/full.wat:23]: I think the description there answers your question, though, unless I've misunderstood it. That is I think an annotation should *not* apply to following instructions, as the example there shows. However an annotation *should* apply to nested children iirc.

**Comment by tlively** [comment]
@kripken, I think this is pretty much ready to go. Do you have opinions on how we should land this switch? Should we make the switch right away, or should we get into a place where we will be able to delete the old parser at the same time? If the former, should we have a `--old-wat-parser` flag for a bit to go back to using the old parser?

**Comment by kripken** [comment]
I don't have a specific idea in mind for how to land this. Either deleting the old parser at the same time, or not, both of those sound acceptable to me. The only critical things I can think of are:

1. Tests pass. (Looks like there's a lit test error atm, is that expected?)
2. Fuzzer passes. That depends on https://github.com/WebAssembly/binaryen/issues/3989 so that we can fuzz the text format again.
3. Coordination with users.
4. Me to review the >100 test file changes here. Before I start on that, why are there so many? I'd hope for 0 :smile: but more realistically I'd expect a few in tests that check corner cases of parsing, but not >100...

**Comment by tlively** [comment]
> I don't have a specific idea in mind for how to land this. Either deleting the old parser at the same time, or not, both of those sound acceptable to me. The only critical things I can think of are:
> 
> 1. Tests pass. (Looks like there's a lit test error atm, is that expected?)

Just pushed a fix for this one :)

> 2. Fuzzer passes. That depends on [Identical function types not merged #3989](https://github.com/WebAssembly/binaryen/issues/3989) so that we can fuzz the text format again.

Cool, I'll look into that.

> 3. Coordination with users.

I can post on the Discord binaryen channel about the switch and also check in with j2wasm.

> 4. Me to review the >100 test file changes here. Before I start on that, why are there so many? I'd hope for 0 😄 but more realistically I'd expect a few in tests that check corner cases of parsing, but not >100...

These are all due to differences in how `try` is parsed and how unreachable code is parsed. It turns out we have a lot of tests with unreachable code!

**Comment by kripken** [comment]
In what way is our parsing of try and unreachable code different? That would also be good to document in the changelog or elsewhere.

**Comment by tlively** [comment]
Check out the changed tests for examples, but in short we explicitly drop anything that was on the stack when we encounter an unreachable instruction, so we end up with more drops than the old parser. For `try`, we often emit a wrapper block with the `try`'s label rather than attaching the label to the `try` itself. I could try making that behavior match the old parser better, but I don't think it's that important.

File: check.py [hunk 3]
```diff
@@ -242,9 +242,9 @@ def check_expected(actual, expected):
                 for module, asserts in support.split_wast(wast):
                     print('        testing split module', split_num)
                     split_num += 1
-                    support.write_wast('split.wast', module, asserts)
+                    support.write_wast('split.wast', module)
                     run_opt_test('split.wast')    # also that our optimizer doesn't break on it
-                    result_wast_file = shared.binary_format_check('split.wast', verify_final_result=False, original_wast=wast)
+                    result_wast_file = shared.binary_format_check('split.wast', verify_final_result=False)
```
Comment (kripken) [inline check.py:247]: These changes look correct and NFC but perhaps they could land in a separate PR?

File: src/parser/wat-parser.cpp [hunk 4]
```diff
@@ -138,7 +138,11 @@ void propagateDebugLocations(Module& wasm) {
       return std::make_unique<Propagator>();
     }
   };
-  PassRunner runner(&wasm);
+  // The parser should not be responsible for validation.
+  PassOptions opts;
+  opts.validate = false;
+  opts.validateGlobally = false;
+  PassRunner runner(&wasm, opts);
```
Comment (kripken) [inline src/parser/wat-parser.cpp:145]: Maybe a better way to get this outcome is to do `setIsNested(true)`. Nested pass runners do not run validation and do not show up on various logging, as they are internal to a pass. (This isn't a pass, but the idea would be the same - it's an internal operation we don't consider as "atomic" in the sense of having valid input and output, and there is no point to log it.)

File: test/lit/basic/polymorphic_stack.wast [hunk 5]
```diff
@@ -219,8 +219,11 @@
   ;; CHECK-TEXT:      (func $untaken-break-should-have-value (type $0) (result i32)
   ;; CHECK-TEXT-NEXT:  (block $x (result i32)
   ;; CHECK-TEXT-NEXT:   (block
-  ;; CHECK-TEXT-NEXT:    (br_if $x
+  ;; CHECK-TEXT-NEXT:    (drop
   ;; CHECK-TEXT-NEXT:     (i32.const 0)
+  ;; CHECK-TEXT-NEXT:    )
+  ;; CHECK-TEXT-NEXT:    (br_if $x
+  ;; CHECK-TEXT-NEXT:     (unreachable)
   ;; CHECK-TEXT-NEXT:     (unreachable)
```
Comment (kripken) [inline test/lit/basic/polymorphic_stack.wast:227]: This change may have implications for our testing. Before, we were able to make a particular argument to this `br_if` be unreachable, and test specifically the proper handling of that, while after, it looks like all arguments become unreachable. My concern is that unreachability of the first argument may prevent testing of handling of only the second argument being unreachable (which is possible after optimizations).

There is also a minor downside to having less precise roundtripping here when debugging.

How difficult would it be to make the parser roundtrip more accurately here?

Comment (tlively) [reply test/lit/basic/polymorphic_stack.wast:227]: Consider these two instruction sequences:

```
i32.const 0
unreachable
i32.add

f32.const 0
unreachable
i32.add
```

They are both valid, and we currently parse both just fine:

```
(drop
 (i32.const 0)
)
(i32.add
 (unreachable)
 (unreachable)
)

(drop
 (f32.const 0)
)
(i32.add
 (unreachable)
 (unreachable)
)
```

If we tried to match existing behavior with the first sequence, we would end up with this for the second sequence:

```
(i32.add
 (f32.const 0)
 (unreachable)
)
```

Taking a look at our validator, it looks like this might be ok, actually? I don't know how consistent we are about letting children of unreachable nodes have any type.

(Also our validator will fail to reject `(i32.add (unreachable) (f32.const 0))`, but that's a separate issue.)

Comment (kripken) [reply test/lit/basic/polymorphic_stack.wast:227]: Hmm, I think that would error - the parent still requires the non-unreachable children make sense.

I agree this is an issue for the unfolded form, but it seems that it has a downside for the folded form. And we use that folded form in order to test our IR, so being able to express less there is concerning.

Comment (tlively) [reply test/lit/basic/polymorphic_stack.wast:227]: After talking offline, the plan is to try augmenting `IRBuilder` with some additional smarts around the expected types of things it pops. This will let the parser continue parsing instructions without having to worry about their nested relationships.

Today when an unreachable instruction is pushed, everything before it is automatically dropped. We will instead defer that automatic dropping behavior as far as possible. When popping the children for each newly-parsed instruction, we will attempt to pop existing instructions from the stack, even past an unreachable instruction. If we pop past an unreachable and encounter a mismatch between the expected and available type, we will undo the popping back to and including the unreachable, then fall back to the auto-drop behavior we have today. We will also auto-drop everything when finishing a scope.

I will implement this after getting fuzzing of text round-tripping working with the current implementation. I can probably extend subtype-exprs.h to pass a pointer to the child expression to `noteSubtype` and build the necessary functionality off of that. We could also reuse this utility in the validator.

File: test/lit/debug/full.wat (hunk 1 above)
Comment (vouillon) [reply test/lit/debug/full.wat]: This might be a bug in the current parser, but currently annotations do apply to the following instructions. For instance, try running the following module through `wasm-opt`:
```
(module
   (func $f
      ;;@ a:1:1
      (nop)
      (nop)
      (nop)
   )
)
````
I get the following ouput:
```
(module
 (type $0 (func))
 (func $f
  ;;@ a:1:1
  (nop)
  ;;@ a:1:1
  (nop)
  ;;@ a:1:1
  (nop)
 )
)
```

Comment (tlively) [reply test/lit/debug/full.wat]: Under the assumption that the old parser is working as intended, the new parser now matches that behavior as well. It's a little unclear to me whether that is the best behavior, though.

File: test/lit/metadce/sourcemap.wat [hunk 6]
```diff
@@ -1,26 +1,34 @@
...
      ;;@ a:3:1
   )
 )
-;; CHECK:      (type $0 (func))
+;; BIN:      (type $0 (func))
 
-;; CHECK:      (export "f" (func $0))
+;; BIN:      (export "f" (func $0))
 
-;; CHECK:      ;;@ a:1:1
```
Comment (vouillon) [inline test/lit/metadce/sourcemap.wat:21]: You are no longer checking that the prologue location is preserved.

Comment (tlively) [reply test/lit/metadce/sourcemap.wat]: Yep, the test auto updater doesn't know about prologue source comments yet, so it's a trade off between having test coverage for them and making the test manual.

File: src/binaryen-c.cpp [hunk 7]
```diff
@@ -1320,17 +1320,12 @@ BinaryenExpressionRef BinaryenBinary(BinaryenModuleRef module,
...
   auto* ret = ((Module*)module)->allocator.alloc<Select>();
   ret->condition = (Expression*)condition;
   ret->ifTrue = (Expression*)ifTrue;
   ret->ifFalse = (Expression*)ifFalse;
-  if (type != BinaryenTypeAuto()) {
-    ret->finalize(Type(type));
-  } else {
-    ret->finalize();
-  }
+  ret->finalize();
```
Comment (kripken) [inline src/binaryen-c.cpp:1328]: Looks like the PR contains changes from other PRs?

Comment (tlively) [reply src/binaryen-c.cpp]: Restacked onto #6389, so this should be fixed.

File: src/parser/wat-parser.cpp (hunk 4 above)
Comment (kripken) [reply src/parser/wat-parser.cpp:145]: I see a thumbs up here but the code looks unchanged - is this in progress or did you prefer the current form for some reason?

Comment (tlively) [reply src/parser/wat-parser.cpp:145]: Sorry, graphite keeps marking this ready for review, but it's definitely still a draft. I'll make this change before this is ready for real review, but I've been focusing on fixing fuzz bugs before coming back to this.

**Comment by tlively** [comment]
With all the recent fixes, `RoundtripText` has run for 100k+ iterations and hasn't found a new bug yet. I've started work on a `ChildTyper` utility that determines what type child expressions should have based on information in their parent. Hopefully we will be able to use it in IRBuilder, the validator, and Unsubtyping. Once it's ready and proven, would you be willing to help modify the validator to use it?

**Comment by kripken** [comment]
Sure, happy to help!

File: test/lit/passes/global-effects.wast [hunk 8]
```diff
@@ -423,7 +423,7 @@
   ;; WITHOUT-NEXT:  )
   ;; WITHOUT-NEXT: )
   ;; INCLUDE:      (func $call-throw-or-unreachable-and-catch (type $2) (param $x i32)
-  ;; INCLUDE-NEXT:  (try $try
+  ;; INCLUDE-NEXT:  (try
```
Comment (kripken) [inline test/lit/passes/global-effects.wast:426]: The amount of changes in this PR is very large. Is it worth trying to reduce diffs like this in preparatory PRs?

If not, can you detail the reasons for the various categories of differences? Some were puzzling to me. I can write specific questions if you prefer.

Comment (tlively) [reply test/lit/passes/global-effects.wast:426]: The categories I'm seeing are:

1. Elem and data segments get slightly different generated names.
2. Loops no longer get generated names if there are no branches.
3. Source mappings are no longer automatically applied to epilogues.
4. Empty then and else clauses are parsed as nops instead of empty blocks.
5. Some tests are fixed to properly use strings in export clauses.
6. Parse error tests are updated.
7. Exports are parsed in a slightly different order.
8. Some tests are updated to respect import ordering constraints.
9. Tests of pops inside of blocks are deleted since they no longer parse. @aheejin thought this would be ok.
10. Old pre-1.0 spec tests that use nonstandard syntax are removed.
11. The tests for parsing Poppy IR directly are removed.

and of course the big one:

12. Try expressions are parsed with wrapper blocks and their labels are replaced with \_\_delegate\_\_ labels.

I'll think about whether the fixing the last one would be feasible since it accounts for the vast majority of hard-to-understand diff-verbosity. I'll also try to pick some low-hanging fruit for the other ones, but I think at some point minimizing the diff more will no longer be worth the extra complexity.

**Comment by tlively** [comment]
The test failure is due to the fact that wasm-shell still uses the old parser while wasm-opt uses the new text parser, so import-after-func.fail.wast does _not_ fail as expected when run under wasm-shell, but then it fails unexpectedly later when run with wasm-opt.

I'll try to fix it by getting wasm-shell able to run with the new text parser in a preliminary PR, since we want to do that anyway.

File: test/lit/passes/global-effects.wast (hunk 8 above)
Comment (tlively) [reply test/lit/passes/global-effects.wast:426]: After #6505, the number of lines added in the diff went down by about 2/3.

File: src/parser/wat-parser.cpp (hunk 4 above)
Comment (kripken) [reply src/parser/wat-parser.cpp:145]: I still think this is worth doing.

File: test/lit/debug/full.wat [hunk 9]
```diff
@@ -46,7 +45,6 @@
   ;; FULL-NEXT:    (i32.const 2)
   ;; FULL-NEXT:   )
   ;; FULL-NEXT:  ) ;; end block block
-  ;; FULL-NEXT:  ;;@ src.cpp:1:2
```
Comment (kripken) [inline test/lit/debug/full.wat:49]: Was this incorrect before?

Comment (tlively) [reply test/lit/debug/full.wat:49]: Unclear what behavior is "correct," but we no longer propagate debug locations to the function epilogue automatically.

**Comment by tlively** [comment]
I am marking this as non-draft for real now. @kripken, PTAL!

File: test/lit/debug/full.wat (hunk 9 above)
Comment (kripken) [reply test/lit/debug/full.wat:49]: I would verify that no emscripten debug info tests fail with this change. If that's ok then sgtm.

File: test/passes/remove-unused-names_merge-blocks_all-features.txt [hunk 10]
```diff
@@ -166,14 +166,11 @@
...
-    (drop
-     (i32.const 10)
-    )
-    (unreachable)
-   )
+  (drop
+   (i32.const 10)
   )
+  (unreachable)
+  (return)
```
Comment (kripken) [inline test/passes/remove-unused-names_merge-blocks_all-features.txt:173]: This looks like a loss of the ability to test such a nested block in a return - is this fixable?

Comment (tlively) [reply test/passes/remove-unused-names_merge-blocks_all-features.txt:173]: The surrounding function has no return values, so the `return` has no business having children. In the old parser we allowed this anyway if the children happened to be unreachable, but I actually think the new behavior makes more sense. Perhaps we should add a check to our validator that only returns in functions that have return values can have children.

Comment (kripken) [reply test/passes/remove-unused-names_merge-blocks_all-features.txt:173]: Thanks, I missed it had no returns from the function. This does sound better then, I agree. Validator improvement makes sense too.

File: test/lit/passes/type-merging.wast [hunk 11]
```diff
@@ -1052,21 +1052,23 @@
...
-  ;; CHECK-NEXT:   (drop
-  ;; CHECK-NEXT:    (br_on_cast $__binaryen_fake_return (ref $A) (ref $B)
-  ;; CHECK-NEXT:     (local.get $a)
-  ;; CHECK-NEXT:    )
-  ;; CHECK-NEXT:   )
-  ;; CHECK-NEXT:   (drop
-  ;; CHECK-NEXT:    (block $l (result (ref $A))
-  ;; CHECK-NEXT:     (br_on_non_null $l
+  ;; CHECK-NEXT:  (block $label (result (ref $B))
+  ;; CHECK-NEXT:   (block (result (ref $B))
```
Comment (kripken) [inline test/lit/passes/type-merging.wast:1056]: It looks like now we generate two blocks here instead of one. This is probably fine as optimizations will remove the extra one, but just wanted to check this is expected? The thing I'd worry about most here is if extra blocks are added in each roundtrip here.

Comment (tlively) [reply test/lit/passes/type-merging.wast:1056]: The inner block is the original function body block (which normally would not be printed) and the outer one is a wrapper block for receiving the branch. It looks like the old parser combines these into a single block, which is slightly nicer. There's no risk of roundtrips adding more blocks here since the branches no longer target the function scope.

Comment (kripken) [reply test/lit/passes/type-merging.wast:1056]: I see, thanks. Would it be hard to reuse the existing function body block in the new parser, as a minor optimization here? Not that important but it would be nice.

File: test/lit/passes/catch-pop-fixup-eh-old.wast
Comment (kripken) [inline test/lit/passes/catch-pop-fixup-eh-old.wast:1]: Why did this file go away?

File: test/lit/passes/type-merging.wast (hunk 11 above)
Comment (tlively) [reply test/lit/passes/type-merging.wast:1056]: I can look into it 👍

File: test/lit/passes/catch-pop-fixup-eh-old.wast
Comment (tlively) [reply test/lit/passes/catch-pop-fixup-eh-old.wast:1]: The new parser cannot parser `pop` except at the correct location at the beginning of the block, but this test file was full of tests that had it elsewhere. I talked to @aheejin about this and she thought deleting this file would be fine since we have other coverage for the fixup pass.

**Review by kripken — APPROVED** [review]
Great!
